Add option of bounding box for PolygonSelector

dstansby opened the pull request:
## PR Summary
This allows one to put a bounding box around a PolygonSelector, that can be used for resizing and moving (and eventually rotating) the selector points.

```python
import matplotlib.pyplot as plt
from matplotlib.widgets import PolygonSelector

fig, ax = plt.subplots()
sel = PolygonSelector(ax, lambda *args, **kwargs: None, draw_box=True)
plt.show()
```
https://user-images.githubusercontent.com/6197628/144757647-ca03dba5-2ab4-4812-a9dd-e80b294b286e.mov

## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->
**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [x] New features are documented, with examples if plot related.
- [x] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [x] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of main, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

dstansby commented:
This should be ready for review now

ericpre commented on lib/matplotlib/widgets.py:
```diff
@@ -3443,6 +3457,66 @@ def __init__(self, ax, onselect, useblit=False,
+        return self._selection_artist.get_bbox()
+
+    def _add_box(self):
+        self._box = RectangleSelector(self.ax,
+                                      onselect=self._scale_polygon,
+                                      useblit=self.useblit,
+                                      interactive=True)
```
Change it on move, instead of onselect:
```suggestion
        self._box = RectangleSelector(self.ax,
                                      onselect=lambda *args, **kwargs: None,
                                      useblit=self.useblit,
                                      interactive=True)
        self._box.connect_event('motion_notify_event', self._scale_polygon)
```
`self._scale_polygon` will need to be changed accordingly.

dstansby replied:
Ah amazing, I was wondering how to do this!

ericpre commented on lib/matplotlib/widgets.py:
```diff
@@ -2795,6 +2795,11 @@ def __init__(self, ax, onselect, drawtype='box',
+        # State to allow the option of an interactive selector that can't be
+        # interactively drawn. This is used in PolygonSelector as an
+        # interactive bounding box to allow the polygon to be easily resized
+        self._allow_creation = True
```
Can `self._selection_completed` used instead?

dstansby replied:
I tried that initially and it didn't work - there's a couple of places in the logic that I changed below that needed this new property.

ericpre commented on lib/matplotlib/widgets.py:
```diff
@@ -3408,7 +3421,8 @@ class PolygonSelector(_SelectorWidget):
     def __init__(self, ax, onselect, useblit=False,
-                 props=None, handle_props=None, grab_range=10):
+                 props=None, handle_props=None, grab_range=10,
+                 draw_box=False):
```
Rename `draw_box` to `draw_bbox`?

ericpre replied:
Add a `bbox_props` and `bbox_handle_props` to pass to `RectangleSelector`?

dstansby replied:
Since this is user facing, I think `box` instead of `bbox` is better - most users won't have come across the concept of bounding boxes or the abbreviation `bbox`.

Good idea to pass through kwargs to RectangleSelector, I will do that.

ericpre commented on lib/matplotlib/widgets.py:
```diff
@@ -3443,6 +3457,66 @@ def __init__(self, ax, onselect, useblit=False,
+    def _remove_box(self):
+        if self._box is not None:
+            self._box.set_visible(False)
+            del self._box
```
Something I am not sure: is it necessary to delete before setting `self._box` to `None`?

dstansby replied:
Maybe not, I will check...

dstansby replied:
No, it wasn't necessary, thanks for catching!

ericpre reviewed (commented):
This works nicely, I made a couple of comments/suggestions.

dhomeier commented on lib/matplotlib/widgets.py:
```diff
@@ -3443,6 +3465,71 @@ def __init__(self, ax, onselect, useblit=False,
+                                      handle_props=self._box_handle_props,
+                                      props=self._box_props,
+                                      interactive=True)
+        self._box.connect_event('motion_notify_event', self._scale_polygon)
+        self._update_box()
```
```suggestion
        self._box._state_modifier_keys.pop('rotate')
        self._box.connect_event('motion_notify_event', self._scale_polygon)
        self._update_box()
```
Currently the box with the default modifiers responds to the `r` key and can be rotated, but of course without effect on the polygon. Until rotation support is added as well, I think this could avoid some confusion.

dhomeier reviewed (changes requested):
Thanks for the PR and updates; I have tested this with a method in astropy/regions#406 that uses this new functionality and see that some early quirks have already been resolved!
A general issue with that `regions` method, which connects a custom `PolygonSelector`, is that creating a complete instance is somewhat tedious – adding the vertices and then triggering a redraw and finally marking the selector `_selection_completed`; for most of this I have not found public methods for the required operations. Having those, analogously to `extents` etc. in `RectangleSelector` would be a plus (but not necessary from this PR).

ericpre commented on lib/matplotlib/tests/test_widgets.py:
```diff
@@ -1464,6 +1478,45 @@ def onselect(vertices):
+    do_event(tool._box, 'press', xdata=40, ydata=40, button=1)
+    do_event(tool._box, 'onmove', xdata=20, ydata=20, button=1)
+    do_event(tool._box, 'release', xdata=20, ydata=20, button=1)
+    assert tool.verts == [(10, 0), (0, 10), (10, 20), (20, 10)]
```
This is failing because the `motion_notify_event` is not triggered, I think?

dstansby replied:
I think that's right. I'm struggling to check how I can get it to trigger though. Because `do_event` goes directly to the widget `onmove`, it doesn't trigger any canvas events. I've tried
```python
    mock_event(ax, button='down', xdata=40, ydata=40, step=0)
    mock_event(ax, button=1, xdata=20, ydata=20, step=0)
    mock_event(ax, button='up', xdata=20, ydata=20, step=0)
```
instead, but that doesn't seem to be working either (the box itself doesn't even resize with this).

ericpre replied:
I don't know if this is the right way to doing it but adding `ax.figure.canvas.motion_notify_event(0, 0)` is working.

dstansby replied:
This and associated methods works, thanks for the tip.

ericpre reviewed (approved):
This looks good, it just need the test to be fixed!

QuLogic commented on lib/matplotlib/widgets.py:
```diff
@@ -3579,7 +3599,8 @@ class PolygonSelector(_SelectorWidget):
     def __init__(self, ax, onselect, useblit=False,
-                 props=None, handle_props=None, grab_range=10):
+                 props=None, handle_props=None, grab_range=10, *,
+                 draw_box=False, box_handle_props={}, box_props={}):
```
It seems okay here, but you should avoid mutable defaults.

dstansby replied:
Thanks for catching - I'll default these to `None` instead.

QuLogic commented on lib/matplotlib/widgets.py:
```diff
@@ -3615,6 +3636,73 @@ def __init__(self, ax, onselect, useblit=False,
         self.set_visible(True)
+        self._draw_box = draw_box
+        self._box = None
+        self._box_handle_props = (
+            self._handle_props.update(box_handle_props) or None)
+        self._box_props = box_props or None
```
Also, this falls back to `None`, which would trigger for `{}`, so why not default to `None` in the signature?

QuLogic commented on lib/matplotlib/widgets.py:
```diff
@@ -3615,6 +3636,73 @@ def __init__(self, ax, onselect, useblit=False,
+        # Create transform from old box to new box
+        x1, y1, w1, h1 = self._box._rect_bbox
+        old_bbox = self._get_bbox()
+        t = transforms.Affine2D()
+        t = t.translate(-old_bbox.x0, -old_bbox.y0)
+        t = t.scale(1 / old_bbox.width, 1 / old_bbox.height)
+        t = t.scale(w1, h1)
+        t = t.translate(x1, y1)
```
```suggestion
        t = (transforms.Affine2D()
             .translate(-old_bbox.x0, -old_bbox.y0)
             .scale(1 / old_bbox.width, 1 / old_bbox.height)
             .scale(w1, h1)
             .translate(x1, y1))
```

QuLogic reviewed (commented):
I wonder if it should be called `draw_bbox`, or even `draw_bounding_box` (since we seem to have pretty verbose names in widgets already.)

dstansby commented:
Thanks for reviewing. I changed the argument name to `draw_bounding_box`, but left the internal variables as `_draw_box` for simplicity. All other comments should be addressed. I also cleaned up the commit history a bit - sorry if this makes it a bit trickier to re-review...

dstansby commented on lib/matplotlib/tests/test_widgets.py:
```diff
@@ -1464,6 +1478,56 @@ def onselect(vertices):
+def test_polygon_selector_box():
+    # Create a diamond shape
+    verts = [(20, 0), (0, 20), (20, 40), (40, 20)]
+    event_sequence = (polygon_place_vertex(*verts[0]) +
+                      polygon_place_vertex(*verts[1]) +
+                      polygon_place_vertex(*verts[2]) +
+                      polygon_place_vertex(*verts[3]) +
+                      polygon_place_vertex(*verts[0]))
```
I didn't do this since the rest of `test_widgets.py` uses the `+` notation. I'd be happy to put in a PR to change it all in one go though?

QuLogic commented on doc/users/next_whats_new/polygon_selector_box.rst:
```diff
@@ -0,0 +1,6 @@
+PolygonSelector bounding boxes
+------------------------------
+`~matplotlib.widgets.PolygonSelector` now has a *draw_box* argument, which
```
Has been renamed.
```suggestion
`~matplotlib.widgets.PolygonSelector` now has a *draw_bounding_box* argument, which
```

QuLogic commented on examples/widgets/polygon_selector_demo.py:
```diff
@@ -50,7 +50,7 @@ def __init__(self, ax, collection, alpha_other=0.3):
             self.fc = np.tile(self.fc, (self.Npts, 1))
 
-        self.poly = PolygonSelector(ax, self.onselect)
+        self.poly = PolygonSelector(ax, self.onselect, draw_box=True)
```
```suggestion
        self.poly = PolygonSelector(ax, self.onselect, draw_bounding_box=True)
```
